### ISSUE TITLE
Support for parallelisation of analysis (utilise several CPUs)

### ISSUE THREAD

**Issue by shakaran** [PHP-CS-Fixer/PHP-CS-Fixer#2803]
I am noticing in big php projects that run some filters will take hours (mostly when a lot config arrays are in place, for example projects that include other libs inside his source code, like tcpdf in older versions) that takes more than some hours in run the fixers. I notice that it is only using one CPU and I cannot run in parallel, unless that I split several instances in several files or folders at same time.

I would like know if there are a way that the project could accept run in parallel for several cpus (for example one CPU could take the fixer for one file and other CPU for other), so if you have 4 CPUs with 8 threads, you can "fix" 8 files per execution, more quick indeed

**Comment by keradus** [comment]
Hi !
There is no such possibility yet, you are more than welcome to propose solution that would bring that feature !

In the meantime, I suggest to ensure that you are running the newest fixer under PHP 7+ with caching enabled.

**Comment by shakaran** [comment]
@keradus thanks! Probably php-cs-fixer could adopt in some way after the param fetching the pcntl_fork(). It duplicates the current PHP process into a child, which can then work separately from the parent process. So per process could init a parsing or rules or fixers.
I am actually using PHP 7.1 with caching. I notice a missing feature about show-progress, I will fill another issue for that.

**Comment by keradus** [comment]
I would prefer to use threads instead of processes, using pool pattern and have solution that would provide common interface and hide parallel-implementation thus we rely on same interface regardless parallelization is possible or not.

**Comment by GrahamCampbell** [comment]
> I would prefer to use threads instead of processes, using pool pattern and have solution that would provide common interface and hide parallel-implementation thus we rely on same interface regardless parallelization is possible or not.

👍 For threads. By parallel implementation, what we really mean is, "concurrent" implementation. :trollface: Just pointing out this, for when naming of classes happens.

**Comment by keradus** [comment]
are any of you willing to create POC? if not, I do really doubt that it will happen soon

**Comment by shakaran** [comment]
:+1: threads seems more suitable.

I think that the POC could be built around the [FixCommand](https://github.com/shakaran/PHP-CS-Fixer/blob/master/src/Console/Command/FixCommand.php#L187) just before to launch the "[Runner](https://github.com/shakaran/PHP-CS-Fixer/blob/master/src/Runner/Runner.php)" class with all the arguments fetched.

Probably the Runner class could be in fact a [thread](http://php.net/manual/en/class.thread.php) or inherit form a Worker/Thread customized class. So a method like "Divide and Conquer" for "attack" all the files and folders in a directory given (and stop when a file parse is unique as stop condition). So a maximum_threads could be added in configuration per default.

**Comment by vertexvaar** [comment]
Since `pcntl_fork` is [not enabled in php by default](http://php.net/manual/en/pcntl.installation.php) and [real multi-threading](http://www.php.net/manual/en/book.pthreads.php) is mostly never installed i would suggest using `proc_open` which can spawn other processes which run concurrently. The nicest and easiest way to achieve this is using [symfony/process](https://github.com/symfony/process). As a working example i would bring (JakubOnderka/PHP-Parallel-Lint)[https://github.com/JakubOnderka/PHP-Parallel-Lint] into this. I you are interested in a simple task manager: I have contributed long time ago to a project which acts as a process manager: [jagandecapri/symfony-parallel-process](https://github.com/jagandecapri/symfony-parallel-process).

I hope this helps :wink:

**Comment by badmansan** [comment]
Any updates about this feature?

**Comment by Wirone** [comment]
@keradus if we have waited 10 days more with the release, this issue would have been solved exactly on its 7th birthday 😂.

**Comment by keradus** [comment]
shall i revert? :P
